### PR TITLE
WT-7984 Fix a bug that could cause a checkpoint to omit a page of data

### PR DESCRIPTION
**Pull request by michaelcahill**
Here are the steps involved:

1. page P exists on disk with address A and is clean
2. checkpoint starts running
3. page P is modified, setting first_dirty_txn ahead of the checkpoint
4. eviction chooses P to evict (in some tree ahead of the checkpoint)
5. eviction reconciles P
6. the main part of reconciliation succeeds but _rec_hs_wrapup fails
   with EBUSY
7. at this point, ref->addr == NULL && mod->rec_result == 0 and the
   block for A has been freed, the page is dirty but first_dirty_txn is
   ahead of the checkpoint
8. checkpoint skips writing P, and when it writes P’s parent, it
   considers P, sees the missing address and takes the WT_CHILD_IGNORE
   path — i.e., nothing is written and the original content of P (from
   step 1) is missing from the checkpoint

The fix is to reorder __rec_write_wrapup to call __rec_hs_wrapup before it clears out the address.

### CODE REVIEW

**Comment by keithbostic** [comment]
@michaelcahill, could we assert reconciliation never fails once the page has been "written", that is, once the address has been updated? Maybe that would prevent this happening again, and we'd panic if it happens in the field.

File: src/reconcile/rec_write.c [hunk 1]
```diff
@@ -218,8 +218,11 @@ __reconcile(WT_SESSION_IMPL *session, WT_REF *ref, WT_SALVAGE_COOKIE *salvage, u
     /* Wrap up the page reconciliation. */
     if (ret == 0 && (ret = __rec_write_wrapup(session, r, page)) == 0)
         __rec_write_page_status(session, r);
-    else
+    else {
+        /* Make sure that reconciliation doesn't free the page that has been written to disk. */
+        WT_ASSERT(session, page->dsk == NULL || ref->addr != NULL);
```
Comment (kommiharibabu) [inline src/reconcile/rec_write.c:223]: @michaelcahill, @keithbostic I added the above assert to ensure that the in-memory page has the disk address and ref doesn't. It is possible to clear the ref->addr in the page delete scenario, but that reconciliation shouldn't fail.

Comment (kommiharibabu) [reply src/reconcile/rec_write.c:223]: The assert fired in our PR testing, Will look into it and update.

Comment (kommiharibabu) [reply src/reconcile/rec_write.c]: I changed the assert as if there exists an address and after wrapup, it should have the address also in the error scenario.

**Comment by michaelcahill** [comment]
@keithbostic, I agree that we should do what we can to avoid future variations on this bug.  After some discussion, @kommiharibabu added an assertion -- please take a look and comment on whether you're happy with that.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Pull request merged**